### PR TITLE
Remove Python 3.4 backports from utils

### PR DESCRIPTION
**Pull request by gmarkall**
Utils contained a couple of backported items from Python 3.4 that are no longer necessary:

- `weakref.finalize`
- `functools.total_ordering`

This PR deletes the backports and replaces uses with the standard library versions. There were a couple of small related changes:

- Whilst testing, I noticed that `test_cuda_no_boundscheck` can fail if you have a CUDA device, if it gets run by a child process - I've added a commit to make sure it has the `SerialMixin` to prevent this.
- The mechanism for ensuring that `shutting_down` is set before `finalizer` exitfuncs run (added in PR #2024) is slightly modified, because this was implemented by modifying `finalize`, which is no longer applicable. I believe my logic for the change is sound, and I haven't noticed and problems (either in general running, or running things and hitting Ctrl-C to interrupt them an random times) but a second person's thoughts on the logic (https://github.com/numba/numba/compare/master...gmarkall:grm-remove-backports?expand=1#diff-e8ce4dc3f72e453a103a52be6032ff8bR293) would probably be good :-)

### CODE REVIEW

**Comment by gmarkall** [comment]
Removing [WIP] now that CI has passed.

File: numba/utils.py [hunk 1]
```diff
@@ -291,6 +290,16 @@ def shutting_down(globals=globals):
...
 
 
+# weakref.finalize registers an exit function that runs all finalizers for
+# which atexit is True. Some of these finalizers may call shutting_down() to
+# check whether the interpreter is shutting down. For this to behave correctly,
+# we need to make sure that _at_shutdown is called before the finalizer exit
+# function. Since atexit operates as a LIFO stack, we first contruct a dummy
+# finalizer then register atexit to ensure this ordering.
+weakref.finalize(lambda: None, lambda: None)
+atexit.register(_at_shutdown)
```
Comment (stuartarchibald) [inline numba/utils.py]: I looked in the CPython source... I think:
1. ` weakref.finalize(lambda: None, lambda: None)` will create and initialise a new finalizer object, in the process of initialisation it will register its own classmethod for invoking finalizers with `atexit`. If a `finalize` object has already been created elsewhere this mechanism will already have run and an internal flag in the `weakref.finalize` impl will prevent re-running. Therefore either the above or some other call to `weakref.finalize` will ensure the classmethod for invoking finalizers is registered with `atexit` by this point in the import sequence.
2. `atexit.register(_at_shutdown)` will register the `_at_shutdown` function with `atexit`, as the `atexit` stack is LIFO, this will be executed ahead of the `weakref.finalize` class method registered in 1. This means that `_at_shutdown` will always have run ahead of any finalizer that may need information about this state via a call to `shutting_down()`.

Basically, this seems correct!

Comment (gmarkall) [reply numba/utils.py]: Thanks - that mirrors my thinking (and also I reached the same conclusion via the same method)!

**Comment by seibert** [comment]
internal CI: `numba_smoketest_cuda_21`

**Comment by stuartarchibald** [comment]
Assuming CI passes I think this can go in 0.48.0.

**Comment by seibert** [comment]
internal CI has passed

**Comment by stuartarchibald** [comment]
> internal CI has passed

Great, thanks.
